### PR TITLE
400 add LrScheduleHandler for engines

### LINKED ISSUE #400: Add LrScheduleHandler for engines
**Is your feature request related to a problem? Please describe.**
When developing a training program based on ignite, we need LrScheduleHandler to update the learning rate when "epoch completed".

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #400 .

### Description
This PR implemented the LrScheduleHandler to adjust the learning rate during training.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/handlers/lr_schedule_handler.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+        """
+        self.lr_scheduler = lr_scheduler
+
+    def attach(self, engine):
+        engine.add_event_handler(Events.EPOCH_COMPLETED, self)
+
+    def __call__(self, engine):
+        self.lr_scheduler.step()
+        print(f"Update learning rate to: {self.lr_scheduler.get_last_lr()[0]}")
```
Comment (ericspod) [inline monai/handlers/lr_schedule_handler.py:34]: This should be a log message rather than printing to stdout.

File: monai/handlers/lr_schedule_handler.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ignite.engine import Events
+
+
+class LrScheduleHander:
```
Comment (ericspod) [inline monai/handlers/lr_schedule_handler.py:15]: LrScheduleHander -> LrScheduleHandler

File: monai/handlers/lr_schedule_handler.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                lr_scheduler object. if customized version, must have `step` and `get_last_lr` methods.
+
+        """
+        self.lr_scheduler = lr_scheduler
+
+    def attach(self, engine):
+        engine.add_event_handler(Events.EPOCH_COMPLETED, self)
+
+    def __call__(self, engine):
+        self.lr_scheduler.step()
```
Comment (ericspod) [inline monai/handlers/lr_schedule_handler.py:33]: For most LR schedulers this works but ones such as [ReduceLROnPlateau](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau) `step` takes arguments. We could add a callable value to `LrScheduleHandler` which if not None is called here and is expected to return positional and keyword arguments to be expanded into the `step` call.

File: monai/handlers/lr_schedule_handler.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        """
+        Args:
+            lr_scheduler (torch.optim.lr_scheduler): typically, lr_scheduler should be PyTorch
+                lr_scheduler object. if customized version, must have `step` and `get_last_lr` methods.
+
+        """
+        self.lr_scheduler = lr_scheduler
+
+    def attach(self, engine):
+        engine.add_event_handler(Events.EPOCH_COMPLETED, self)
```
Comment (ericspod) [inline monai/handlers/lr_schedule_handler.py:30]: Are there not schedulers which must be stepped after every iteration?

File: tests/test_handler_lr_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            strides=(2, 2, 2, 2),
+            num_res_units=2,
+        )
+        optimizer = torch.optim.Adam(net.parameters(), 0.1)
+        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1, gamma=0.1)
+        handler = LrScheduleHander(lr_scheduler)
+        handler.attach(engine)
+
+        engine.run(data, max_epochs=5)
+        np.testing.assert_allclose(lr_scheduler.get_last_lr()[0], 1e-06, rtol=1e-06)
```
Comment (ericspod) [inline tests/test_handler_lr_scheduler.py:45]: Perhaps we can assert that the learning rate for the optimizer has decayed by the expected amount? If so we should use SGD instead of an adaptive optimizer like Adam so that we see this more clearly.

File: monai/handlers/lr_schedule_handler.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/handlers/lr_schedule_handler.py:34]: updated in the latest patch.
Thanks.

File: monai/handlers/lr_schedule_handler.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/handlers/lr_schedule_handler.py:15]: updated in the latest patch.
Thanks for correcting my typo.

File: tests/test_handler_lr_scheduler.py (hunk 5 above)
Comment (Nic-Ma) [reply tests/test_handler_lr_scheduler.py:45]: updated to SGD in the latest patch.
Thanks.

File: monai/handlers/lr_schedule_handler.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/handlers/lr_schedule_handler.py:30]: updated to support every epoch or every iteration in the latest patch.
Thanks.

File: monai/handlers/lr_schedule_handler.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/handlers/lr_schedule_handler.py:33]: Thanks for your reminder.
I added callable function and updated unit test in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review, I updated the PR according to your comments.
Could you please help review it again?
Thanks.

File: tests/test_handler_lr_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        val_engine = Engine(_train_func)
+        train_engine = Engine(_train_func)
+
+        @train_engine.on(Events.EPOCH_COMPLETED)
+        def run_validation(engine):
+            val_engine.run(data)
+            val_engine.state.metrics["val_loss"] = 1
+
+        # set up testing handler
+        net = UNet(
```
Comment (ericspod) [inline tests/test_handler_lr_scheduler.py:40]: If you wanted to simplify things `net` could just be a single nn.Module object like conv2d if you just need something to pass to the optimizer.

File: monai/handlers/lr_schedule_handler.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                `True` is epoch level, `False` is iteration level.
+            step_transform (Callable): a callable that is used to transform the information from `engine`
+                to expected input data of lr_scheduler.step() function if neccessary.
+
+        """
+        self.lr_scheduler = lr_scheduler
+        self.print_lr = print_lr
+        self.logger = None if name is None else logging.getLogger(name)
+        self.epoch_level = epoch_level
+        assert isinstance(step_transform, Callable), "step_transform must be a callable function."
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/handlers/lr_schedule_handler.py:40]: updated in the latest patch.
Thanks.

File: monai/handlers/lr_schedule_handler.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    def attach(self, engine):
+        if self.logger is None:
+            self.logger = engine.logger
+        if self.epoch_level:
+            engine.add_event_handler(Events.EPOCH_COMPLETED, self)
+        else:
+            engine.add_event_handler(Events.ITERATION_COMPLETED, self)
+
+    def __call__(self, engine):
+        args = ensure_tuple(self.step_transform(engine))
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/handlers/lr_schedule_handler.py:52]: I checked the PyTorch lr_schedulers, the base class uses `step(epoch=None)`, so I set the default as None.
Anyway, I changed to `lambda engine: ()` in the latest patch.
Thanks.

File: tests/test_handler_lr_scheduler.py (hunk 6 above)
Comment (Nic-Ma) [reply tests/test_handler_lr_scheduler.py:40]: updated to use `torch.nn.PReLU()` in the latest patch.
Thanks.

File: tests/test_handler_lr_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+            dimensions=3,
+            in_channels=1,
+            out_channels=1,
+            channels=(16, 32, 64, 128, 256),
+            strides=(2, 2, 2, 2),
+            num_res_units=2,
+        )
+        optimizer = torch.optim.SGD(net.parameters(), 0.1)
+        lr_scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=1)
+        handler = LrScheduleHandler(lr_scheduler, step_transform=lambda x: val_engine.state.metrics["val_loss"])
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply tests/test_handler_lr_scheduler.py:50]: added another test in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review, I updated the PR again according to your comments.
Could you please help review it again?
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

I slightly updated the unit test to make it more clear and extensible.
Thanks.
